collectors/debugfs.plugin/debugfs_plugin.c

```diff
@@ -0,0 +1,224 @@
+        if(!freq) {
+            int n = (int)str2l(argv[i]);
+            if(n > 0) {
+                freq = n;
+                continue;
+            }
+        }
+
+        if(strcmp("test-permissions", argv[i]) == 0 || strcmp("-t", argv[i]) == 0) {
+            if(!debugfs_check_sys_permission()) {
+                perror("Cannot open /sys/kernel/debug/extfrag/extfrag_index file");
+                exit(1);
+            }
+            printf("OK\n");
+            exit(0);
+        }
+    }
+
+    if(freq > 0) update_every = freq;
+}
+
+int main(int argc, char **argv)
+{
+    // debug_flags = D_PROCFILE;
+    stderror = stderr;
+
+    // set the name for logging
+    program_name = "debugfs.plugin";
+
+    // disable syslog for debugfs.plugin
+    error_log_syslog = 0;
+
+    netdata_configured_host_prefix = getenv("NETDATA_HOST_PREFIX");
+    if (verify_netdata_host_prefix() == -1)
+        exit(1);
+
+    user_config_dir = getenv("NETDATA_USER_CONFIG_DIR");
+    if (user_config_dir == NULL) {
+        user_config_dir = CONFIG_DIR;
+    }
+
+    stock_config_dir = getenv("NETDATA_STOCK_CONFIG_DIR");
+    if (stock_config_dir == NULL) {
+        // info("NETDATA_CONFIG_DIR is not passed from netdata");
+        stock_config_dir = LIBCONFIG_DIR;
+    }
+
+    if (!debugfs_check_capabilities() && !debugfs_am_i_running_as_root()) {
+        uid_t uid = getuid(), euid = geteuid();
+#ifdef HAVE_CAPABILITY
+        error(
+            "debugfs.plugin should either run as root (now running with uid %u, euid %u) or have special capabilities. "
+            "Without these, debugfs.plugin cannot access /sys/kernel/debug. "
+            "To enable capabilities run: sudo setcap cap_dac_read_search,cap_sys_ptrace+ep %s; "
+            "To enable setuid to root run: sudo chown root:netdata %s; sudo chmod 4750 %s; ",
+            uid,
+            euid,
+            argv[0],
+            argv[0],
+            argv[0]);
+#else
+        error(
+            "debugfs.plugin should either run as root (now running with uid %u, euid %u) or have special capabilities. "
+            "Without these, debugfs.plugin cannot access /sys/kernel/debug."
+            "Your system does not support capabilities. "
+            "To enable setuid to root run: sudo chown root:netdata %s; sudo chmod 4750 %s; ",
+            uid,
+            euid,
+            argv[0],
+            argv[0]);
+#endif
+    }
+
+    debugfs_parse_args(argc, argv);
+
+    size_t iteration;
+    usec_t step = update_every * USEC_PER_SEC;
+    heartbeat_t hb;
+    heartbeat_init(&hb);
+    for (iteration = 0; iteration < 86400; iteration++) {
+        heartbeat_next(&hb, step);
+
+        int i;
+        for (i = 0; debugfs_modules[i].name; i++) {
+            struct debugfs_module *pm = &debugfs_modules[i];
+            if (unlikely(!pm->enabled))
+                continue;
+
+            pm->enabled = !pm->func(update_every, pm->name);
+        }
+    }
+
+    return 0;
```

Comment:
debugfs never exits, is that expected?

Reply:
It exits when receives `SIGTERM` from netdata. This is default behavior of other external plugins like [cups.plugin](https://github.com/netdata/netdata/blob/89568295b62c65f848dbee013a0d6677a0aa24e9/collectors/cups.plugin/cups_plugin.c#L258).

Reply:
Yes, but there is no modules enabled?

Reply:
When I ran on a VM without swap I still had metrics for it, but they were always zero, so it looks like that  the unique possibility o have everything disabled is when `debugfs` is not mounted. Last commit addressed this.